extras/dht/dht.c

```diff
@@ -0,0 +1,141 @@
+ *  Note:
+ *  A suitable pull-up resistor should be connected to the selected GPIO line
+ *
+ *  __           ______          _______                              ___________________________
+ *    \    A    /      \   C    /       \   DHT duration_data_low    /                           \
+ *     \_______/   B    \______/    D    \__________________________/   DHT duration_data_high    \__
+ *
+ *
+ *  Initializing communications with the DHT requires four 'phases' as follows:
+ * 
+ *  Phase A - MCU pulls signal low for at least 18000 us
+ *  Phase B - MCU allows signal to float back up and waits 20-40us for DHT to pull it low
+ *  Phase C - DHT pulls signal low for ~80us
+ *  Phase D - DHT lets signal float back up for ~80us
+ *  
+ *  After this, the DHT transmits its first bit by holding the signal low for 50us
+ *  and then letting it float back high for a period of time that depends on the data bit.
+ *  duration_data_high is shorter than 50us for a logic '0' and longer than 50us for logic '1'.
+ *
+ *  There are a total of 40 data bits trasnmitted sequentially. These bits are read into a byte array
+ *  of length 5.  The first and third bytes are humidity (%) and temperature (C), respectively.  Bytes 2 and 4
+ *  are zero-filled and the fifth is a checksum such that:
+ *
+ *  byte_5 == (byte_1 + byte_2 + byte_3 + btye_4) & 0xFF
+ *
+*/
+
+/*  
+ *  @pin                    the selected GPIO pin
+ *  @interval               how frequently the pin state is checked in microseconds
+ *  @timeout                maximum length of time to wait for the expected pin state
+ *  @expected_pin_state     high (true) or low (false) pin state
+ *  @counter                pointer to external uint8_t for tallying the duration waited for the pin state
+*/
+
+bool dht_await_pin_state(uint8_t pin, uint8_t interval, uint8_t timeout, bool expected_pin_state, uint8_t * counter) {
+
+    for (*counter = 0; *counter < timeout; *counter+=interval) {
+        if (gpio_read(pin) == expected_pin_state) return true;
+        sdk_os_delay_us(interval);
+    }
+
+    return false;
+}
+
+/*  
+ *  
+ *  
+ *  @pin                    the selected GPIO pin
+ *  @internal_pullup        boolean to indicate whether the internal pullup resistor should be used
+ *  @humidity               pointer to external int8_t to store resulting humidity value
+ *  @temperature            pointer to external int8_t to store resulting temperature value
+ *  @status_code            pointer to external uint8_t to store status code corresponding to the following messages:
+ *                              0: "Success!"
+ *                              1: "Checksum failed, invalid data received from sensor..."
+ *                              2: "DHT did not initialize correctly, problem in phase 'B'..."
+ *                              3: "DHT did not initialize correctly, problem in phase 'C'..."
+ *                              4: "DHT did not initialize correctly, problem in phase 'D'..."

```

Comment:
If these constants are part of the API, it would be better to give them symbolic names in `dht.h` (using either `#define` or by using an `enum` datatype).  As I mentioned in another comment, though, the more I think about this, the more I think it's probably not necessary to return this info to the caller at all, and these could just be debugging `printf` calls, etc, instead (which would simplify the API too)..  Not a huge deal either way, though..